### DISCUSSION TITLE
Cypress tests failing in GitLab CICD Pipeline due to Environment variables.

### DISCUSSION THREAD

**Discussion by Sakshi3110** [cypress-io/cypress#14203] [CI setup] [answered]
cc : @jennifer-shehane  @brian-mann 
 **I have added env in cypress.json like in below format -** 
"env": {
        "qa": {
          "baseURL": "https://www.xyz.com",
                     "challengeID": "4567",
        },
        "staging": {
            "baseURL": "https://www.xyzaa.com",
                 "challengeID": "1234",
        }
**My test spec files have :** 
                           const env = Cypress.env('env');
                           const baseLink = Cypress.env(env).baseURL;
                          	cy.visit(`${baseLink}`);

On local everything runs fine, as I provide --env env=qa in command CLI.
**But in Git, there is a global generic code. npx run --config baseURL= 'www.xyz.com'** , so we cannot edit it to add env in command.
In git templates we have environments defined for stages in this way :
 before_script:
     - export pipelineId=${CI_PIPELINE_ID}
    - export GITLAB_TOKEN=${GITLAB_TOKEN}
    - export TEST_TYPE=cypress
    **- export ENV=devint**
We should read env variable from process.env.ENV , that ENV variable is OS based variable. Its is generic way for all kinds of test not just cypress. 

So, I tried changing my code and in test files I wrote : 
const env = Cypress.env('process.env.ENV');
const env = Cypress.env('process.env.NODE_ENV')
const env = Cypress.env('ENV');
Nothing worked.

**const env = Cypress.env('qa');**  This works perfectly.
 But I want to make it generic , I am not sure how to capture env in code.
I don't want to hardcode env value qa/devint in code. It will be good if we can make it generic and capture ENV from git only.
Can you please guide how can we use already given ENV variable in code ?

Thanks and regards,
Sakshi

**Comment by bahmutov** [comment, marked as answer]
If you want to load and set different environment variables based on `process.env` values, you need to use the plugins file https://on.cypress.io/environment-variables#Option-5-Plugins

We have a recipe that shows several ways of setting and passing environment variables here https://github.com/cypress-io/cypress-example-recipes/tree/master/examples/server-communication__env-variables

And I have a section in my "Cypress tips and tricks" blog post specifically describing an example https://glebbahmutov.com/blog/cypress-tips-and-tricks/#pass-the-environment-variables-correctly

In your case the `cypress/plugins/index.js` file would have something like

```js
module.exports = (on, config) => {
  const envName = process.env.ENV || process.env.NODE_ENV || 'qa'
  console.log('using env %s', envName)

  // we want to use baseUrl depending on the env name
  const baseUrl = config.env[envName].baseURL

  if (baseUrl) {
    console.log('setting base URL to %s', baseUrl)
    config.baseUrl = baseUrl
  }  

  // return config with changed variables
  return config
}
```

**Reply by Sakshi3110** [reply]
Thank you so much @bahmutov for such a quick response.
I referred to your suggestion and made below changes in plugin file. I also made few changes in test specs & json files and called envName variable in test files. 
```
//index.js
module.exports = (on, config) => {
const envName = process.env.ENV
config.env[envName]
return config
}
```
```
//spec.js - I made changes in this file also
const env = Cypress.config('envName');
const baseURL = Cypress.env(env).baseurl; //this calls env which is described in cypress.json.
```
```
//cypress.json
"env": {
		"qa": {
			"baseURL": "http://www.abc.com",
			"commonUI": "http://www.xyz.com/"
		},
		"devint": {
				"baseURL": "http://www.abc1.com",
			"commonUI": "http://www.xyz1.com/"
		},
```
Now all my tests are passing successfully in git pipelines, for all the environments. 💯 ✌️
